File: src/framework/operator.cpp
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "framework/operator.h"
#include <memory>
#include "operators/op_param.h"
namespace paddle_mobile {
namespace framework {

template <typename Dtype>
vector<string> OperatorBase<Dtype>::GetOutKeys() const {
  auto it = op_input_output_key.find(type_);
  if (it == op_input_output_key.end()) {
    DLOG << type_ << " has no outputs";
    return {};
  }
  return it->second.second;
}

template <typename Dtype>
vector<string> OperatorBase<Dtype>::GetInputKeys() const {
  auto it = op_input_output_key.find(type_);
  if (it == op_input_output_key.end()) {
    DLOG << type_ << " has no inputs";
    return {};
  }
  return it->second.first;
}

template <typename Dtype>
OperatorBase<Dtype>::OperatorBase(const std::string &type,
                                  const VariableNameMap &inputs,
                                  const VariableNameMap &outputs,
                                  const AttributeMap &attrs,
                                  framework::Scope *scope)
    : type_(type),
      inputs_(inputs),
      outputs_(outputs),
      attrs_(attrs),
      scope_(scope) {
  CheckAllInputOutputSet();
}

template <typename Dtype>
void OperatorBase<Dtype>::CheckAllInputOutputSet() const {}

template <typename Dtype>
void OperatorBase<Dtype>::Run() {
#ifdef PADDLE_MOBILE_DEBUG
  static int index_input = 0;
  DLOG << "[" << index_input << "] ------------ " << type_ << " ------------";
  for (const auto key : GetInputKeys()) {
    DLOG << "input key: " << key;
    auto var_vec_in = inputs_.at(key);
    for (int i = 0; i < var_vec_in.size(); ++i) {
      auto var = this->scope_->FindVar(var_vec_in[i]);
      if (var->IsInitialized() &&
          var->template IsType<framework::LoDTensor>()) {
        const Tensor *tensor = var->template Get<framework::LoDTensor>();
        if (tensor)
          DLOG << "\ttensor name: " << var_vec_in[i] << " " << *tensor;
// #ifdef PADDLE_MOBILE_FPGA_KD
//         if (tensor) {
//           std::string path = "input/" + std::to_string(index_input) +
//                              "_input__" + var_vec_in[i] + "__" +
//                              tensor->zynqmpTensor()->dimsFileName();
//           // DLOG << "\tfile name: " << path.c_str();
//           tensor->zynqmpTensor()->readFromFile(path);
//         }
// #endif
#ifdef PADDLE_MOBILE_FPGA
        DLOG << var_vec_in[i];
#endif
      }
    }
  }
  index_input++;
#endif
  RunImpl();
#ifdef PADDLE_MOBILE_DEBUG
  static int index_output = 0;
  for (const auto key : GetOutKeys()) {
    auto var_vec_out = outputs_.at(key);
    DLOG << "output key: " << key;
    for (int i = 0; i < var_vec_out.size(); ++i) {
      auto var = scope_->FindVar(var_vec_out[i]);
      if (var->IsInitialized() &&
          var->template IsType<framework::LoDTensor>()) {
        const Tensor *tensor = var->template Get<framework::LoDTensor>();
        if (tensor)
          DLOG << "\ttensor name: " << var_vec_out[i] << " " << *tensor;
// #ifdef PADDLE_MOBILE_FPGA_KD
//         if (tensor) {
//           std::string path = "output/" + std::to_string(index_output) +
//                              "_output__" + var_vec_out[i] + "__" +
//                              tensor->zynqmpTensor()->dimsFileName();
//           // DLOG << "\tfile name: " << path.c_str();
//           tensor->zynqmpTensor()->save_file_with_name(path);
//         }
// #endif
#ifdef PADDLE_MOBILE_FPGA
        DLOG << var_vec_out[i];
#endif
      }
      //       if (var->IsInitialized() &&
      //           var->template IsType<std::vector<framework::LoDTensor>>()) {
      //         const std::vector<framework::LoDTensor> *vec =
      //             var->template Get<std::vector<framework::LoDTensor>>();
      //         DLOG << "\ttensor name: " << var_vec_out[i] << " " <<
      //         (*vec)[0];
      // #ifdef PADDLE_MOBILE_FPGA_KD
      //         std::string path = "output/" + std::to_string(index_output) +
      //                            "_output__" + var_vec_out[i] + "__" +
      //                            (*vec)[0].zynqmpTensor()->dimsFileName();
      //         // DLOG << "file name: " << path.c_str();
      //         (*vec)[0].zynqmpTensor()->save_file_with_name(path);
      // #endif
      //       }
    }
  }
  index_output++;
#endif
}

#ifdef PADDLE_MOBILE_CL
template <>
void OperatorBase<GPU_CL>::Run() {
  RunImpl();
#ifdef PADDLE_MOBILE_DEBUG
  DLOG << "-------------" << type_ << "----------------------------";
  vector<string> input_keys = GetInputKeys();
  for (const auto key : input_keys) {
    auto var_vec_in = inputs_.at(key);
    for (int i = 0; i < var_vec_in.size(); ++i) {
      auto var = scope_->FindVar(var_vec_in[i]);
      if (var->IsInitialized() && var->template IsType<framework::CLImage>()) {
        const CLImage *cl_image = var->template Get<framework::CLImage>();
        if (cl_image) {
          DLOG << type_ << " input- " << key << "=" << *cl_image;
        }
      }
    }
  }
  for (const auto key : GetOutKeys()) {
    auto var_vec_out = outputs_.at(key);
    for (int i = 0; i < var_vec_out.size(); ++i) {
      auto var = scope_->FindVar(var_vec_out[i]);
      if (var->IsInitialized() && var->template IsType<framework::CLImage>()) {
        const CLImage *cl_image = var->template Get<framework::CLImage>();
        if (cl_image) {
          DLOG << type_ << " output- " << key << "=" << *cl_image;
        }
      }
    }
  }
#endif
}
#endif

#ifdef PADDLE_MOBILE_FPGA
template <typename Dtype>
void OperatorBase<Dtype>::InsertTensors() {
  static int feed_num = 0;
  static int fetch_num = 0;
  if (type_ == "feed") {
    auto new_name = string("feed") + std::to_string(feed_num++);
    auto var = scope_->Var(new_name);
    var->template GetMutable<framework::LoDTensor>();
    inputs_.at("X") = {string(new_name)};
  } else if (type_ == "fetch") {
    auto new_name = string("fetch") + std::to_string(fetch_num++);
    auto var = scope_->Var(new_name);
    var->template GetMutable<framework::LoDTensor>();
    outputs_.at("Out") = {string(new_name)};
  }
}
#endif

template class OperatorBase<CPU>;
template class OperatorBase<FPGA>;
template class OperatorBase<GPU_MALI>;
template class OperatorBase<GPU_CL>;

}  // namespace framework
}  // namespace paddle_mobile

File: metal/paddle-mobile/paddle-mobile/Operators/Operator.swift
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License. */

import Foundation

/*
 let opInputsOutputsKey  = [gConvType         : (inputs: ["Input"], outputs: ["Output"]),
 gBatchNormType    : (inputs: ["X"], outputs: ["Y"]),
 gReluType         : (inputs: ["X"], outputs: ["Out"]),
 gElementwiseAdd   : (inputs: ["X", "Y"], outputs: ["Out"])]
 */

protocol Param {
    associatedtype ParamP: PrecisionType
    init(opDesc: OpDesc, scope: Scope) throws
    static func getFirstTensor(key: String, map: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func inputX(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func inputBiase(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func inputMean(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func inputScale(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func inputVariance(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func inputFilter(paraInputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func input(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func output(outputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func outputY(outputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func inputY(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func outputOut(outputs: [String : [String]], from: Scope) throws -> Tensor<ParamP>
    static func getAttr<T>(key: String, attrs: [String : Attr]) throws -> T
}

extension Param {
    static func getFirstTensor(key: String, map: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        guard let mapKeys = map["X"], mapKeys.count > 0, let inputX = from[mapKeys[0]], let tensorX = inputX as? Tensor<ParamP> else {
            throw PaddleMobileError.paramError(message: "tensor " + key + "in \(map) not found")
        }
        return tensorX
    }
    
    static func inputX(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorX = try getFirstTensor(key: "X", map: inputs, from: from)
            return tensorX
        } catch let error {
            throw error
        }
    }
    
    static func input(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorInput = try getFirstTensor(key: "Input", map: inputs, from: from)
            return tensorInput
        } catch let error {
            throw error
        }
    }
    
    static func output(outputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorOutput = try getFirstTensor(key: "Output", map: outputs, from: from)
            return tensorOutput
        } catch let error {
            throw error
        }
    }
    static func outputY(outputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorOutputY = try getFirstTensor(key: "Y", map: outputs, from: from)
            return tensorOutputY
        } catch let error {
            throw error
        }
    }
    static func inputY(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorY = try getFirstTensor(key: "Y", map: inputs, from: from)
            return tensorY
        } catch let error {
            throw error
        }
    }
    
    static func outputOut(outputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let out = try getFirstTensor(key: "Out", map: outputs, from: from)
            return out
        } catch let error {
            throw error
        }
    }
    static func inputFilter(paraInputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorFilter = try getFirstTensor(key: "Filter", map: paraInputs, from: from)
            return tensorFilter
        } catch let error {
            throw error
        }
    }
    
    static func inputBiase(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorBias = try getFirstTensor(key: "Bias", map: inputs, from: from)
            return tensorBias
        } catch let error {
            throw error
        }
    }

    static func inputMean(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorMean = try getFirstTensor(key: "Mean", map: inputs, from: from)
            return tensorMean
        } catch let error {
            throw error
        }
    }
    
    static func inputScale(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorScale = try getFirstTensor(key: "Scale", map: inputs, from: from)
            return tensorScale
        } catch let error {
            throw error
        }
    }
    
    static func inputVariance(inputs: [String : [String]], from: Scope) throws -> Tensor<ParamP> {
        do {
            let tensorVariance = try getFirstTensor(key: "Variance", map: inputs, from: from)
            return tensorVariance
        } catch let error {
            throw error
        }
    }
    
    static func getAttr<T>(key: String, attrs: [String : Attr]) throws -> T{
        guard let attr = attrs[key] as? T else {
            throw PaddleMobileError.paramError(message: "attr type error")
        }
        return attr
    }
}


class Operator<ParamType: Param> {
    let type: String
    let inputs: [String : [String]]
    let paraInputs: [String : [String]]
    let outpus: [String : [String]]
    let attrs: [String : Attr]
    let para: ParamType
    init(opDesc: OpDesc, inScope: Scope) throws {
        type = opDesc.type
        inputs = opDesc.inputs
        outpus = opDesc.outputs
        attrs =  opDesc.attrs
        paraInputs = opDesc.paraInputs
        do {
            para = try ParamType.init(opDesc:opDesc, scope: inScope)
        } catch let error {
            throw error
        }
    }
    
    func run() {
        runImpl()
    }
    
    func runImpl() {
        fatalError("runimpl of " + type + "op not implement")
    }
}
